### ISSUE TITLE
Allow Serving "Virtual" Routes from Plugins

### ISSUE THREAD

**Issue by illBeRoy** [docsifyjs/docsify#1737]
## Feature request
Plugins are a great way to introduce dynamic content into existing pages. But, what if we took this a step further, and allowed plugins to actually generate the routes themselves on the fly?

The feature I want to contribute introduces an API for plugins to actually provide dynamically generated pages for routes, even if actual markdown files do not exist to begin with.

### What problem does this feature solve?
Say that you have a list of APIs, which can change at any time. Each one should have its own page under `/apis/<METHOD_NAME>.md`.

Today, you will have to create real markdown files and deploy them along with your site, as every route in a Docsify website requires that a file exist on the server. This means that you will have to create all the `/apis/*.md` files manually and deploy them, and do this every time that even a character changes in any of the API descriptions.

With my proposed solution, you will be able to create a plugin that intercepts all fetches from `/apis/*`, and provide dynamically generated markdown as replacement from fetching real markdown files from the server.

### What does the proposed API look like?
Add a new hook to the Plugin API, called: `fetchRoute(cb)`, which will essentially allow plugins to intercept fetch requests before they are made by the underlying engine. If any plugin has a fetchRoute hook, and the callback provided any content for the route, use that content instead of fetching.

#### The fetchRoute Hook Behavior
This hook receives a callback, possibly async, that implements the following signature: `(route: Route, next: fn) => void | Promise<void>`. The route object is the same one that can be found on `vm.route`.

The `next` function can either receive a `markdown` string, or nothing at all. If a non-empty string was provided, then this is the markdown that will be passed into the system, and no `fetch` will be done by the Fetch class itself.

If the `next` function was called with a falsy value, it will just call the next plugin that has `fetchRoute`. Finally, if all plugins were exhausted and none returned any markdown, the internal `fetch` function will be run like today.

#### Flow Diagram
![image](https://user-images.githubusercontent.com/6681893/151674863-5bf49160-e0b9-4038-9eab-3612696bc4d6.png)

![image](https://user-images.githubusercontent.com/6681893/151675085-d7cff35d-207b-43f8-9730-624e8fbf5e85.png)

![image](https://user-images.githubusercontent.com/6681893/151675043-0cce6b73-bfc1-4480-8811-c075e7f68526.png)

### How should this be implemented in your opinion?
From my brief introduction to the code, I would suggest that we edit the `_fetch` method ([link to code](https://github.com/docsifyjs/docsify/blob/40a5fa8f446d0616df4ea33707a8d146a435530a/src/core/fetch/index.js#L84)), so instead of directly using `get`, it will run the `fetchRoute` hooks first and only fall back to get if none return markdown.

### Are you willing to work on this yourself?
Yes :) Would love to.

**Comment by illBeRoy** [comment]
Hey guys,

is there anyone I can tag in order to get this proposal reviewed?
Maybe @jhildenbiddle ? (based on other issues I've seen?)

Thanks in advance for the feedback :)

**Comment by sy-records** [comment]
cc @docsifyjs/reviewers

**Comment by jhildenbiddle** [comment]
Thanks for taking the time to document the request, @illBeRoy. This looks interesting.

I haven't had time to think this through, but here are a few thoughts off the top of my head and in no particular order:

- The proposed `fetchRoute` hook name feels inconsistent with the existing hook names. To be honest, I don't love the existing hook names and they too lack consistency, so my feedback here is admittedly about remaining "consistent with our inconsistency" versus finding the best name possible. A name like `beforeRequest` feels like it aligns better with the current hook names. It also has the benefit of not implying a request method ([`fetch`](https://developer.mozilla.org/en-US/docs/Web/API/fetch) vs. [`XMLHttpRequest`](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest)), which could otherwise cause confusion since the proposed name contains `fetch` but we're using XMLHttpRequest under the hood.
- The proposed hook is also the logical place to programmatically modify request options. For example, plugins could modify the request URL, headers, if/how credentials are sent, etc. To facilitate this, the `next()` function could receive a modified `route` object in additional to a markdown `string`. The shape of the `route` object should be documented, agnostic from existing web APIs, and does not need to support every `fetch()` or `XMLHttpRequest` option available (IMHO).
- Are there any scenarios where `_fetch()` is used to request content that we would not want to send through the plugin chain? My first concern was with [embedded files](https://docsify.js.org/#/embed-files), but it appears this is not a concern from the limited testing I did. I have not spent a lot of time with part of the codebase so this may be a non-issue, but I thought it was worth mentioning.

While outside the scope of this proposal, it's easy to imagine how creating dynamic site content would also be useful for other docsify content such as the top navigation, sidebar, and embeds. Multiple feature requests have been made [here](https://github.com/docsifyjs/docsify/issues/984) and in [other repos](https://github.com/jhildenbiddle/docsify-tabs/issues/31) to allow markdown features to just work everywhere rather than just the main content area. Obviously, docsify isn't setup to do that today and doing so would require significant changes to our plugin system, but it's worth considering as we explore how to integrate the proposed feature. For example, how could a `beforeRequest`-like plugin be used to dynamically generate sidebar content? How about an [embed](https://docsify.js.org/#/embed-files) so pages could be made up of both static and dynamic content?

Exciting stuff. Looking forward to hearing everyone's thoughts. :)

**Comment by trusktr** [comment]
I think `fetchRoute` is a better name than `beforeRequest`. They imply entirely different things. `beforeRequest` implies that it runs *before* a request, which is not what we want, and also begs the question "what about `afterRequest`.

I propose we do not introduce new names that follow bad naming of existing hooks.

Also I don't think `fetchRoute` is confusing. The plugin author can fetch any way they want, whether with `fetch` or not. It's just a verb.

Naming ideas:

- fetchRoute
- getRoute
- requestRoute
- onRoute
- provideRouteContent
- provideRoute

Nite that someone doesn't even have to fetch anything, they could just provide a hard coded markdown string for example.

> My first concern was with embedded files, 

I propose let's not worry about this yet. A first version can allow a plugin to provide markdown for the top level route only, not all other requests. I believe this is what the OP is about too.

After that, embedded files would work the same.

Later on, as a separate feature, onRequest or onResponse hooks could be used to intercept and, if desired by a plugin author, replace the responses or requests with alternates.

**Comment by trusktr** [comment]
Or maybe we just need onResponse and onRequest? That would be more powerful, and we can include metadata as to whether a request or response is for a route or not.

**Comment by jhildenbiddle** [comment]
> @trusktr - I think `fetchRoute` is a better name than `beforeRequest`. They imply entirely different things. `beforeRequest` implies that it runs before a request, which is not what we want...

Running before a request is the proposed behavior. From the OC:

> @illBeRoy - Add a new hook to the Plugin API, called: `fetchRoute(cb)`, which will essentially **_allow plugins to intercept fetch requests before they are made by the underlying engine_**.

There are many reasons why a site owner or plugin author may want to intercept a request. We shouldn't assume we know what that reason is. The goal may have nothing to do with "fetching" data or providing a "route" response at all, which is why I don't think we should name the hook based on that particular use case (or any other, for that matter).

Instead of focusing on providing virtual routes, I think we should approach this feature request as a way of providing the ability to "do stuff" before the system makes a request. The OP's use case -- requesting data from an API and generating content from it -- is one example of how this functionality could be used, but there are countless others; displaying a UI element, prompting for user input/credentials, interacting with a Vue component, changing the state of a remote application, or modifying a request before it is passed on to the system are just a few examples. This is why I'm proposing naming the hook based on state like `beforeRequest` rather than an assumed use case like `fetchRoutes`, `provideRouteContent`, `setRequestHeaders`, etc.

> @jhildenbiddle - My first concern was with embedded files
> @trusktr - I propose let's not worry about this yet.

To clarify, the OP mentioned modifying the existing [`_fetch()`](https://github.com/docsifyjs/docsify/blob/40a5fa8f446d0616df4ea33707a8d146a435530a/src/core/fetch/index.js#L84) method. My comment was about making sure we understand how this method is used so we don't unintentionally feed the wrong type of fetched data (like embeds) into the plugin chain.

I agree that we can provide similar functionality to other areas of docsify (top nav, sidebar, and embeds) down the road. My intention in mentioning how useful this could be even though it was "outside the scope of this proposal" was to encourage @illBeRoy to think beyond the use case provided in the OP.

> @trusktr - `beforeRequest` [...] also begs the question "what about afterRequest"?

Yep!

> @trusktr - Or maybe we just need onResponse and onRequest? That would be more powerful...

Yep! :)

Both are useful.

As far as naming goes, if we offer both request and response related hooks (and I think we should), then I agree that `onRequest`/`onResponse` work better than trying to stick with the existing `before`/`after` convention. My `beforeRequest` suggestion was based on the assumption that this proposal was focused on adding _only_ a new request-related hook.

The `onRequest` hook outlined above (as `beforeRequest`) accommodates the OP's use case as well as each of the examples I gave.

The `onResponse` hook could be used to modify content just like `beforeEach` is today, but it could also be used to inspect response-related information like headers and status codes not available with `beforeEach`. We could also provide helper methods that allow for things like initiating a new request at the beginning of docsify's plugin chain to handle certain kinds of errors. We can flush out the details separately, but knowing that a response-related hook is planned will help us better understand and define the request-related hook IMHO.

Also worth mentioning: In the future we can determine if we need both `onResponse` and `beforeEach`, or if a single hook is sufficient since `onResponse` does everything `beforeEach` does and more. Short-term, we can simply add `onResponse` as a non-breaking change and see how it goes. That's all TBD of course, but I wanted to cover the overlap between `onResponse` and `beforeEach`.

> @trusktr - A first version can allow a plugin to provide markdown for the top level route only, not all other requests. I believe this is what the OP is about too.

I'm not sure I understand what you mean by "top level route only". If you mean the route associated with the main content area then I agree.

**Comment by jhildenbiddle** [comment]
Hey @illBeRoy. 

Thanks for the thoughtful breakdown and your willingness to work through the details.

I agree we're talking about two different approaches, both of which address the  route-focused use case you've outlined. My concern is how this all plays out if/when `onRequest` and `onResponse` hooks are implemented. If we agree that these hooks are useful, I think it makes sense to assume they will be so we can work through potential issues and make better short- and long-term decisions. I also think the pros and cons can easily be made to favor either approach, as can the the rational for implementing one hook before the other.

The good news is that I have a proposal that I believe works well for everyone and addresses the issues discussed so far.

Let's dive in...

### 1. Add `routes` configuration option

This is the "higher-level" API for providing dynamic route-based markdown content. Providing this functionality as a configuration option is consistent with docsify's existing route/request-related options like `alias`, `basePath`, and `requestHeaders`.

```js
// Object
window.$docsify = {
  routes: {
    // Basic match w/ return string
    '/foo/': 'Custom Markdown',
    // RegEx match w/ function
    '/bar/(.*)': function(route, matched, done) {
      console.log(`Route match: ${matched[0]}`);
      done();
    }
  }
}
```

#### Benefits over hook-based routes

- Custom routes without requiring a custom plugin
- Route configurations available for inspection and modification
- URL matching handled by docsify

#### Implementation

- `route` argument is a string that reflects changes made by docsify configuration
- `matches` argument is an array of RegEx matches (if applicable)
- Synchronous functions return directly
- Asynchronous functions return using `done()`
- When a route function returns a string of any length, the value is passed to the `hook.response` plugin chain as `responseText` without an accompanying `xhr` object (see below). This accomplishes the following:
  1. Allows custom routes to render an empty content area
  1. Allows handling of response text using `hook.response` or `hook.beforeEach`
  1. Allows distinguishing between `responseText` from a custom route or a request
- When a route function returns an explicit `false`, the route change is aborted and no change is made to the window location.
- When a route function returns no value or any other falsey value, the next available route will be processed. If the current route is the last route defined, the system will make a standard request and proceed as usual.

### 2. Add `request` hook

Discussed previously as `onRequest`.



```js
hook.request(function(url, xhr, next) {
  // ...
});
```

#### Benefits over `routes` configuration option

- [XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest) object available for inspection and modification before request
- Route configuration is protected from inspection and modification

#### Implementation

- `url` argument is a string that reflects changes made by docsify configuration
- `xhr` is the XMLHttpRequest object [creatd by docsify](https://github.com/docsifyjs/docsify/blob/40a5fa8f446d0616df4ea33707a8d146a435530a/src/core/fetch/ajax.js#L15) before it is [sent](https://github.com/docsifyjs/docsify/blob/40a5fa8f446d0616df4ea33707a8d146a435530a/src/core/fetch/ajax.js#L33).

The behavior is similar to custom route functions with a few key differences:

- Asynchronous functions return using `next()`
- When the hook returns no value or a falsey value other than `false` or an empty string, the next `hook.request` function will be invoked. If the current hook is the last hook in the chain, the system will make a standard request and proceed as usual.

### 3. Add `response` hook

Discussed previously as `onResponse`.

```js
hook.response(function(responseText, xhr, next) {
  // ...
});
```

#### Benefits over `beforeEach` hook

- [XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest) object available for response inspection and modification

#### Implementation

- `responseText` argument is a mutable string (`xhr.responseText` is read-only)
- `xhr` is the XMLHttpRequest object [creatd by docsify](https://github.com/docsifyjs/docsify/blob/40a5fa8f446d0616df4ea33707a8d146a435530a/src/core/fetch/ajax.js#L15) after a response has been [received](https://github.com/docsifyjs/docsify/blob/40a5fa8f446d0616df4ea33707a8d146a435530a/src/core/fetch/ajax.js#L65).

The behavior is similar to the `request` hook with a few key differences:

- When a hook returns a string of any length, the value is passed to the `hook.beforeEach` plugin chain as `content`.
- When the hook returns no value or a falsey value other than `false` or an empty string, the next `hook.response` function will be invoked. If the current hook is the last hook in the chain, the value will be passed to the first `b beforeEach` hook and the system will proceed as usual.

### Notes

1. We can remove the "on" prefix from the request/response hook names. It doesn't make the hook names more descriptive and dropping it provides consistency with existing hooks like `hook.init`, `hook.mounted`, and `hook.ready` (all of which *could* be prefixed with `on` but currently are not).
1. Short-term, the `response` hook is the same as `beforeEach` but with access to additional response data (headers, status code, status text, etc). In the future we can revisit the overlap between the `response` and `beforeEach` hooks as well as hook names.
1. The OP originally mentioned passing the `vm.route` object as as the `route` object to custom route functions. `vm.route` data does not reflect route changes made by docsify configuration options like `alias`, `basePath`, and `relativePath`, while the `url` value passed to `request` hook *will* reflect these changes.

   ```js
   window.$docsify = {
     basePath: '/foo/'
   }
   ```

   `vm.route` object

   ```js
   {
     path: "/",
     file: "README.md",
     query: ""
   }
   ```

   `url` value passed to `request` hook:

   ```
   "/foo/README.md"
   ```

   Passing different values to the custom route functions and `request` hooks will likely cause some confusion. This is why I specified that both the `route` and `url` value passed to these functions should reflects changes made by docsify configuration.

### Proposal

1. Implement the `routes` configuration option above as a stand-alone PR. Users and plugin-authors will then be able to add, remove, and modify custom routes by modifying the `routes` configuration object.
1. After merging the `routes` configuration option, implement both the `request` and `response` hooks in a separate PR.
1. :partying_face:

**Comment by jhildenbiddle** [comment]
@illBeRoy --

Checking in. :smile:

**Comment by illBeRoy** [comment]
Hey! Sorry for the delay, seems like I completely missed the previous update!

This looks like a great idea. Gonna start working on the first PR asap.

**Comment by jhildenbiddle** [comment]
@illBeRoy --

Sounds great. Looking forward to it. Happy to lend a hand if needed.

**Comment by illBeRoy** [comment]
Hey again!

I know it's been a while (had some stuff going on and almost no free time on my hands), but I finally got to work on the PR and I'm almost ready to submit!

Before I do, though, I've a few questions about the spec, and would love your feedback:

## Matches in string routes
Regarding the following:
```js
{
  // RegEx match w/ return string and match string
  '/foo/(.*)': '# Route match: $1',
}
```
While it seems like an extremely convenient way to format routes without having to write too much code, I'm a little worried that people might unknowingly trigger this mechanism.

Now, I thought about possible solutions (e.g. more distinct delimiter; a standard templating language like ejs or vm) but IMO that inherent problem is that people who do not expect it might still fall for it.

#### Bottom Line
I already implemented this feature, but maybe we should drop it altogether? If someone wants to format the match into their pages, let them use a function and not a string.

## The "next" function
I was thinking that maybe we can rethink how we handle async functions? That is, if a function is sync, we just use its return value as the contents of the page.

How about we do the same with async functions, instead of making you invoke a `next` function? API seems a little bit more straightforward imo.

Thanks for your time! LMK what you think and I'll adjust the PR accordingly.

**Comment by jhildenbiddle** [comment]
@illBeRoy --

Very excited to hear about your progress! Thanks for devoting the time and energy on this. Very much appreciated. :)

> While it seems like an extremely convenient way to format routes without having to write too much code, I'm a little worried that people might unknowingly trigger this mechanism.

If I understand the concern correctly, the issue is that a string-based route value my contain `$1` which may unintentionally be replaced with the matched portion of the URL, correct? If so, I think that's a fair point and a good catch. I don't have an issue dropping this feature from string-based route values (i.e. `'# Route match: $1'`) in favor of requiring function-based route values for those that need access to the matched string from the URL (as you suggest).

> I was thinking that maybe we can rethink how we handle async functions? That is, if a function is sync, we just use its return value as the contents of the page.

This is how it works today. The [documentation explaining how to write a plugin](https://docsify.js.org/#/write-a-plugin?id=write-a-plugin) doesn't make this very clear. This (among other reasons) is why I rewrote those docs which you can see on the develop branch preview:

https://docsify-preview.vercel.app/#/write-a-plugin

> How about we do the same with async functions, instead of making you invoke a next function? API seems a little bit more straightforward imo.

No argument from me about the need to revisit out plugin system. How async functions are handled is one of the items on that todo list. The one thing to keep in mind is that the current version of Docsify still supports legacy browsers, changing how async plugins are handled would require a major version bump (I assume), and bumping to the next major version is not as simple as it may seem due to Docsify's history. I'm very much trying to avoid opening up a can of worms and diving into that history here, so apologies if that seems vague.

I think the best approach *for now* is to separate the functionality proposed in this PR from tasks like rethinking how we handle async plugins, then revisit the plugin system and decide if/how/when we want to move forward there in a separate issue. I'd love to have that discussion if it's something you're interested in.

Hope this helps.

**Comment by illBeRoy** [comment]
So as far as I understand, our action items are:
1. Drop support for implicit string formatting (`Page: $1`)
2. Keep support for `next`

Will do.

One more question, though: how do you propose that we determine whether the function is sync or async? since with sync functions we don't wait for anything (just use the returned string), but with async functions we need to await `next` to be called

Thanks!

By the way, you can see my draft [here](https://github.com/docsifyjs/docsify/pull/1799) (need to add docs and modify the impl to match what we discussed here)

**Comment by jhildenbiddle** [comment]
@illBeRoy 

> I see that in Lifecycle, we use function.length to determine whether or not the callback expects a next function ([here](https://github.com/docsifyjs/docsify/blob/develop/src/core/init/lifecycle.js#L42)). So I guess I'll go down that path for now.

Yep. The plugin system implementation predates any of the current maintainers. I just worked within the existing pattern when possible to reduce risk and avoid requiring a major version bump.

FYI, I also approved the draft so automated tests can run on the PR. You should see the results reflected on Github.

**Comment by illBeRoy** [comment]
Thanks!
Working on the PR as we speak.

**Comment by illBeRoy** [comment]
Hey! PR is ready for review 🙌

**Comment by sy-records** [comment]
Hi, @illBeRoy @jhildenbiddle
Can I close this issue? Support completed in #1799?
